src/discrete_alloc.rs

```diff
@@ -0,0 +1,271 @@
+use std::alloc::{self, Layout};
+use std::sync;
+
+use nix::sys::mman::ProtFlags;
+
+use crate::helpers::ToU64;
+
+static ALLOCATOR: sync::Mutex<MachineAlloc> = sync::Mutex::new(MachineAlloc::empty());
+
+/// A distinct allocator for the `MiriMachine`, allowing us to manage its
+/// memory separately from that of Miri itself.
+#[derive(Debug)]
+pub struct MachineAlloc {
+    pages: Vec<*mut u8>,
+    huge_allocs: Vec<(*mut u8, usize)>,
+    allocated: Vec<Box<[u8]>>,
+    page_size: usize,
+    enabled: bool,
+}
+
+// SAFETY: We only point to heap-allocated data
+unsafe impl Send for MachineAlloc {}
+
+impl MachineAlloc {
+    // Allocation-related methods
+
+    /// Initializes the allocator with placeholder 4k pages.
+    const fn empty() -> Self {
+        Self {
+            pages: Vec::new(),
+            huge_allocs: Vec::new(),
+            allocated: Vec::new(),
+            page_size: 4096,
+            enabled: false,
+        }
+    }
+
+    /// SAFETY: There must be no existing `MiriAllocBytes`
+    pub unsafe fn enable() {
+        let mut alloc = ALLOCATOR.lock().unwrap();
+        alloc.enabled = true;
+        // This needs to specifically be the system pagesize!
+        alloc.page_size = unsafe {
+            let ret = libc::sysconf(libc::_SC_PAGE_SIZE);
+            if ret > 0 {
+                ret.try_into().unwrap()
+            } else {
+                4096 // fallback
+            }
+        }
+    }
+
+    /// Returns a vector of page addresses managed by the allocator.
+    #[expect(dead_code)]
+    pub fn pages() -> Vec<u64> {
```

Comment:
Why is this `u64`, not `usize`? These are host pages after all, right?

Probably better to remove the function entirely given that it is dead code, so I cannot judge how it will be used.

Reply:
These all are used only for integrating with the ptrace stuff, but I can remove them for now, sure. The reason it was `u64`s is so that we don't need to iteratively call `to_u64()` in the ptrace code on what this returns, as everything there uses `u64`s internally to be the same as the rest of Miri

Reply:
Generally in Miri we should use `u64` for target addresses and `usize` for host addresses.

But with native-libs mode, target==host, so... I guess that blurs the line.

Anyway let's discuss that in the ptrace PR; please remove these functions for now.